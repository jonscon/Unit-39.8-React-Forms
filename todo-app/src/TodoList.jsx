import NewTodoForm from "./NewTodoForm.jsx";
import Todo from "./Todo.jsx";
import { useState } from 'react';

function TodoList() {
    const [todos, setTodos] = useState([]);

    // add a new todo
    const addTodo = newTodo => {
        setTodos(todos => [...todos, newTodo]);
    }

    // update a todo with updatedTask
    const updateTodo = (id, updatedTask) => {
        setTodos(todos => 
            todos.map(todo => 
                todo.id === id ? {...todo, task: updatedTask} : todo
            )
        );
    };

    // delete a todo by id
    const deleteTodo = id => {
        setTodos(todos => todos.filter(todo => todo.id !== id));
    }

    // create array of todos
    const todoComponents = todos.map(todo => (
        <Todo 
            remove={ deleteTodo }
            key={ todo.id }
            id={ todo.id }
            update={ updateTodo }
            task={ todo.task }
        />
    ));

    return (
        <div>
            <NewTodoForm addTodo={addTodo}/>
            <ul>{ todoComponents }</ul>
        </div>
    );
}

export default TodoList;