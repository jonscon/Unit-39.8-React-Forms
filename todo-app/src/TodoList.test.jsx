import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList.jsx";
import {it, expect } from 'vitest';

function addTodo(todoList, task="test task") {
    const taskInput = todoList.getByLabelText("To-Do");
    fireEvent.change(taskInput, { target: { value: task } });
    const submitButton = todoList.getByText("Add");
    fireEvent.click(submitButton);
}

// Smoke Test
it("renders without crashing", function() {
    render(<TodoList />);
});

// Snapshot Test
it("matches snapshot", function() {
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a todo", function() {
    const list = render(<TodoList />);
    addTodo(list);

    // expect form to clear and todo to be on the page
    expect(list.getByLabelText("To-Do")).toHaveValue("");
    expect(list.getByText("test task")).toBeInTheDocument();
    expect(list.getByText("Edit")).toBeInTheDocument();
    expect(list.getByText("X")).toBeInTheDocument();
})

it("can edit a todo", function() {
    const list = render(<TodoList />);
    addTodo(list);

    fireEvent.click(list.getByText("Edit"));
    const editInput = list.getByDisplayValue("test task");
    fireEvent.change(editInput, { target: { value: "boo" } });
    fireEvent.click(list.getByText("Update"));

    // expect only edited task to appear
    expect(list.getByText("boo")).toBeInTheDocument();
    expect(list.queryByText("test task")).not.toBeInTheDocument();
})

it("can delete a todo", function() {
    const list = render(<TodoList />);
    addTodo(list);

    fireEvent.click(list.getByText("X"));

    // expect task to be gone
    expect(list.queryByText("test task")).not.toBeInTheDocument();
})