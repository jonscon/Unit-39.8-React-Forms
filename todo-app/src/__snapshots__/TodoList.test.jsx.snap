// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="todo"
        >
          To-Do
        </label>
        <input
          id="todo"
          name="todo"
          value=""
        />
        <button>
          Add
        </button>
      </form>
    </div>
    <ul />
  </div>
</DocumentFragment>
`;
